File: android/app/src/main/java/com/ahmedkhaled/onroute/service/AnalyticsService.kt
package com.ahmedkhaled.onroute.service

import android.content.Context
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import org.json.JSONObject
import java.util.UUID

object AnalyticsService {
    private const val BASE_URL = "https://backend-navy-iota.vercel.app"
    private const val PREFS_NAME = "onroute_analytics"
    private const val KEY_ANONYMOUS_ID = "anonymous_id"

    private val client = OkHttpClient()
    private var anonymousId: String = ""

    fun init(context: Context) {
        val prefs = context.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE)
        anonymousId = prefs.getString(KEY_ANONYMOUS_ID, null) ?: run {
            val id = UUID.randomUUID().toString()
            prefs.edit().putString(KEY_ANONYMOUS_ID, id).apply()
            id
        }
    }

    fun track(event: String, properties: Map<String, Any> = emptyMap()) {
        if (anonymousId.isEmpty()) return

        CoroutineScope(Dispatchers.IO).launch {
            try {
                val props = JSONObject()
                properties.forEach { (k, v) -> props.put(k, v) }

                val body = JSONObject().apply {
                    put("event", event)
                    put("anonymousId", anonymousId)
                    put("properties", props)
                }

                val request = Request.Builder()
                    .url("$BASE_URL/api/event")
                    .post(body.toString().toRequestBody("application/json".toMediaType()))
                    .build()

                client.newCall(request).execute().close()
            } catch (_: Exception) { }
        }
    }
}

File: ios/Detour/Services/AnalyticsService.swift
import Foundation

final class AnalyticsService {
    static let shared = AnalyticsService()

    let anonymousId: String

    private init() {
        let key = "analyticsAnonymousId"
        if let existing = UserDefaults.standard.string(forKey: key), !existing.isEmpty {
            anonymousId = existing
        } else {
            let id = UUID().uuidString
            UserDefaults.standard.set(id, forKey: key)
            anonymousId = id
        }
    }

    func track(_ event: String, properties: [String: Any] = [:]) {
        let body: [String: Any] = [
            "event": event,
            "anonymousId": anonymousId,
            "properties": properties,
        ]

        guard let url = URL(string: "\(APIService.baseURL)/api/event"),
              let jsonData = try? JSONSerialization.data(withJSONObject: body) else { return }

        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.timeoutInterval = 10

        Task {
            _ = try? await URLSession.shared.data(for: request)
        }
    }
}
